pandas/io/sql.py

```diff
@@ -1316,7 +1319,12 @@ def _harmonize_columns(
                 elif dtype_backend == "numpy" and col_type is float:
                     # floats support NA, can always convert!
                     self.frame[col_name] = df_col.astype(col_type)
-
+                elif (
+                    using_string_dtype()
+                    and is_string_dtype(col_type)
+                    and is_object_dtype(self.frame[col_name])
```

Comment:
And for example a column of all missing values, that might get inferred as float by the DataFrame constructor? In that case maybe rather `not is_string_dtype` instead of `is_object_dtype`? (Or just pass a `copy=False` to astype)

Reply:
It is needed, although admittedly the patterns used here are not very clear.

Without the `is_object_dtype` check, it is possible that a series that is constructed as a StringDtype using pd.NA as the na_value sentinel gets changed to using the `np.nan` sentinel, which fails a good deal of tests

As an alternative, we might be able to infer the `na_value` to use from the executing environment rather than hard-coding np.nan on L1436?

Reply:
Maybe it would be clearer then to use `if dtype_backend == "numpy"` in the if clause? (instead of `is_object_dtype`) I.e. only do this cast to StringDtype in default backend mode. 
That's also what the case just above and below actually use .. (so maybe those could be combined within a single `if dtype_backend == "numpy"` block)

(this also makes me wonder if we should call `_harmonize_columns` at all in case of a pyarrow dtype_backend, because at that point the `self.frame` already has all ArrowDtype columns. But I suppose this method handles some additional things like the `parse_dates` keyword?)

Reply:
I think we need to revert this to using `is_object_dtype` instead of just checking the `dtype_backend`. 

The problem is showing up in the latest CI failure because of the NUMERIC type in postgres, which is akin to the decimal data type in Python. The current SQL code treats that as an `object` dtype, so this branch ends up getting activated and converts that data into an `object` type array, rather than keeping as float

Reply:
Because `col_type` is `object`

```python
>>> from pandas.core.dtypes.common import is_string_dtype
>>> is_string_dtype(object)
True
```

is that not expected?

Reply:
Ah, yes, I constantly forget that `is_string_dtype` is a bit of a misnomer ..

Ideally we have a check that just passes for actual string dtypes, but let's leave that for another time.